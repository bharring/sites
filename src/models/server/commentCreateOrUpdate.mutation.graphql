mutation CommentCreateOrUpdate($input: CommentInput) {
  commentCreateOrUpdate(input: $input) {
    comment {
      id
      entity_type
      entity_id
      body
      created_at
      created_by {
        id
        tenantId
        email
        phone
        firstName
        lastName
        organizations {
          id
          name
          subdomain
          roles
        }
      }
    }
  }
}
