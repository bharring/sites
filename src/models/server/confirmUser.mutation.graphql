mutation ConfirmUser($userConfirmationInput: UserConfirmationInput) {
  userManagement(userConfirmationInput: $userConfirmationInput) {
    ... on UserConfirmationResultSuccess {
      __typename
      userConfirmation {
        code
        username
      }
    }
    ... on CodeMismatchError {
      username
      code
      message
    }
    ... on UserNotFoundError {
      message
      username
    }
    ... on NotAuthorizedException {
      message
      username
    }
  }
}
