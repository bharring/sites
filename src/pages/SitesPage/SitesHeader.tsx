import { IonButton, IonButtons, IonHeader, IonIcon, IonMenuButton, IonTitle, IonToolbar } from '@ionic/react';

export const SitesHeader: React.FC<{ title: string; icon: string; toggleView: () => void }> = ({ title, icon, toggleView }) => {
  return (
    <IonHeader>
      <IonToolbar>
        <IonButtons slot="start">
          <IonMenuButton></IonMenuButton>
        </IonButtons>
        <IonTitle>{title}</IonTitle>
        <IonButtons slot="primary">
          <IonButton data-testid="button" slot="end" onClick={toggleView}>
            <IonIcon icon={icon} />
          </IonButton>
        </IonButtons>
      </IonToolbar>
    </IonHeader>
  );
};

export default SitesHeader;
