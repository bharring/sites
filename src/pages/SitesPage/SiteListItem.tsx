import { IonLabel } from '@ionic/react';

import { Site } from '../../models/server/schema';

export const SiteListItem: React.FC<{ site: Site | null }> = ({ site }) => (
  <IonLabel>
    {site?.name && <h2>{site.name}</h2>}
    {site?.address && (
      <h3>{`${site.address.street || ''}, ${site.address.locality || ''}, ${site.address.region || ''} ${
        site.address.postalCode || ''
      }`}</h3>
    )}
    {/* TODO: <p>{ site | formatDistance:currentLocation }</p> */}
  </IonLabel>
);

export default SiteListItem;
