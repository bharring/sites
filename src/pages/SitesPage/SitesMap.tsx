import { GoogleMap, Marker, useJsApiLoader } from '@react-google-maps/api';
import { useCallback } from 'react';

import { SitesListQuery } from '../../models/server/schema';
import { useErrorToast } from '../../state/useErrorToast';

const currentLocationIcon: google.maps.Symbol = {
  path: 0, // google.maps.SymbolPath.CIRCLE,
  scale: 10,
  fillOpacity: 1,
  strokeWeight: 2,
  fillColor: '#5384ED',
  strokeColor: '#ffffff',
};

export const SitesMap: React.FC<{ center?: google.maps.LatLngLiteral; data?: SitesListQuery }> = ({ center, data }) => {
  const { isLoaded, loadError } = useJsApiLoader({
    googleMapsApiKey: process.env.REACT_APP_GOOGLE_MAPS_KEY as string,
  });

  const onLoad = useCallback((map) => {
    const bounds = data?.sitesList?.items?.reduce(
      (_bounds, item) =>
        item && item?.position?.latitude && item?.position?.longitude
          ? _bounds.extend(new google.maps.LatLng(item.position.latitude, item.position.longitude))
          : _bounds,
      new google.maps.LatLngBounds(),
    );

    bounds && map.fitBounds(bounds);
  }, []);

  useErrorToast(loadError);

  if (isLoaded) {
    return (
      <GoogleMap
        onLoad={onLoad}
        mapContainerStyle={{
          width: '100%',
          height: '100%',
        }}
      >
        {center && <Marker key="currentPosition" position={center} icon={currentLocationIcon}></Marker>}
        {data?.sitesList?.items &&
          data?.sitesList?.items.map(
            (site) =>
              site?.position?.latitude &&
              site.position?.longitude && (
                <Marker key={site?.id} position={{ lat: site.position.latitude, lng: site.position.longitude }}></Marker>
              ),
          )}
      </GoogleMap>
    );
  }

  return <></>;
};

export default SitesMap;
