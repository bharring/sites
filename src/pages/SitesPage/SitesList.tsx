import { RefresherEventDetail } from '@ionic/core';
import { IonItem, IonList, IonRefresher, IonRefresherContent } from '@ionic/react';

import { SitesListQuery } from '../../models/server/schema';
import SiteListItem from './SiteListItem';

const SitesList: React.FC<{
  data?: SitesListQuery;
  refetch: (variables?: any) => Promise<any>; // this is the problem with using generated code, the actual function signature of refetch is unwieldy
}> = ({ data, refetch }) => {
  const handleRefresh = async (e: CustomEvent<RefresherEventDetail>) => {
    try {
      await refetch();
    } catch {}
    e.detail.complete();
  };

  return (
    <>
      <IonRefresher slot="fixed" onIonRefresh={handleRefresh}>
        <IonRefresherContent></IonRefresherContent>
      </IonRefresher>

      <IonList>
        {data?.sitesList?.items &&
          data?.sitesList?.items.map((site) => (
            <IonItem button key={site?.id}>
              <SiteListItem site={site} />
            </IonItem>
          ))}
      </IonList>
    </>
  );
};

export default SitesList;
