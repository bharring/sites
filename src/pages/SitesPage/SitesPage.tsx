import { RefresherEventDetail } from '@ionic/core';
import { IonContent, IonLoading, IonPage, IonRefresher, IonRefresherContent } from '@ionic/react';
import { listOutline, mapOutline } from 'ionicons/icons';
import { useState } from 'react';

import { useSitesListQuery } from '../../models/server/schema';
import { useErrorToast } from '../../state/useErrorToast';
import { useWatchPosition } from '../../state/useGeolocation';
import SitesHeader from './SitesHeader';
import SitesList from './SitesList';
import SitesMap from './SitesMap';

const SitesPage: React.FC = () => {
  const { data, loading, error, refetch } = useSitesListQuery({ notifyOnNetworkStatusChange: true });
  const { currentLatLng, geolocationError } = useWatchPosition();

  useErrorToast(geolocationError);
  useErrorToast(error);

  const [icon, setIcon] = useState(mapOutline);
  const toggleView = () => setIcon(icon === mapOutline ? listOutline : mapOutline);

  return (
    <IonPage>
      <SitesHeader title="Sites" icon={icon} toggleView={toggleView} />

      <IonLoading isOpen={loading}></IonLoading>

      <IonContent>
        {icon === mapOutline ? (
          <SitesList data={data} refetch={refetch}></SitesList>
        ) : (
          <SitesMap center={currentLatLng} data={data}></SitesMap>
        )}
      </IonContent>
    </IonPage>
  );
};

export default SitesPage;
