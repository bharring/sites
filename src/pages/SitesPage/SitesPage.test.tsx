import '@testing-library/jest-dom';

import { MockedProvider } from '@apollo/client/testing';
import { cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { SitesListDocument } from '../../models/server/schema';
import SitesPage from './SitesPage';

const mockSprings = {
  id: '00000000-0000-0000-0000-000000000000',
  name: 'Bartons Springs Pool',
  address: {
    street: '2201 William Barton Dr',
    locality: 'Austin',
    region: 'TX',
    postalCode: '78746',
    country: 'US',
  },
  position: {
    latitude: 30.2641138,
    longitude: -97.7734661,
    altitude: null,
  },
};
const mockCapitol = {
  id: '11111111-1111-1111-1111-111111111111',
  name: 'Texas Capitol',
  address: {
    street: '1100 Congress Ave',
    locality: 'Austin',
    region: 'TX',
    postalCode: '78701',
    country: 'US',
  },
  position: {
    latitude: 30.2746698,
    longitude: -97.7425445,
    altitude: null,
  },
};
describe('<SitesPage />', () => {
  afterEach(cleanup);

  test('matches snapshot', () => {
    const mocks = [
      {
        request: {
          query: SitesListDocument,
        },
        result: {
          data: {
            sitesList: { items: [mockSprings, mockCapitol], nextToken: null, scannedCount: null },
          },
        },
      },
    ];

    const tree = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <SitesPage />
      </MockedProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
});
