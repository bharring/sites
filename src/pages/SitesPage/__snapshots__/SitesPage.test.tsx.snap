// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SitesPage /> matches snapshot 1`] = `
<div
  className="ion-page"
>
  <ion-header>
    <ion-toolbar>
      <ion-buttons
        slot="start"
      >
        <ion-menu-button />
      </ion-buttons>
      <ion-title>
        Sites
      </ion-title>
      <ion-buttons
        slot="primary"
      >
        <ion-button
          data-testid="button"
          onClick={[Function]}
          slot="end"
        >
          <ion-icon
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Map</title><path d='M313.27 124.64L198.73 51.36a32 32 0 00-29.28.35L56.51 127.49A16 16 0 0048 141.63v295.8a16 16 0 0023.49 14.14l97.82-63.79a32 32 0 0129.5-.24l111.86 73a32 32 0 0029.27-.11l115.43-75.94a16 16 0 008.63-14.2V74.57a16 16 0 00-23.49-14.14l-98 63.86a32 32 0 01-29.24.35zM328 128v336M184 48v336' stroke-linecap='round' stroke-linejoin='round' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
          />
        </ion-button>
      </ion-buttons>
    </ion-toolbar>
  </ion-header>
  <ion-content>
    <ion-refresher
      slot="fixed"
    >
      <ion-refresher-content />
    </ion-refresher>
    <ion-list />
  </ion-content>
</div>
`;
